File: server/src/main/java/com/securechat/server/dao/MessageDAO.java
package com.securechat.server.dao;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * MessageDAO – CRUD cho bảng messages (lịch sử chat).
 */
public class MessageDAO {

    private static final Logger log = LoggerFactory.getLogger(MessageDAO.class);

    /**
     * Lưu tin nhắn mã hoá vào database.
     *
     * @param roomId           ID phòng chat
     * @param senderId         ID người gửi
     * @param encryptedContent nội dung đã mã hoá AES (Base64)
     * @param type             TEXT hoặc FILE
     * @param fileName         tên file (null nếu TEXT)
     * @param fileSize         kích thước file (0 nếu TEXT)
     * @return ID của message vừa lưu, -1 nếu lỗi
     */
    public static int saveMessage(int roomId, int senderId,
                                   String encryptedContent, String type,
                                   String fileName, long fileSize) {
        String sql = "INSERT INTO messages " +
                     "(room_id, sender_id, encrypted_content, iv, message_type, file_name, file_size) " +
                     "VALUES (?, ?, ?, '', ?, ?, ?)";

        try (Connection conn = DatabaseConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {

            ps.setInt(1, roomId);
            ps.setInt(2, senderId);
            ps.setString(3, encryptedContent);
            ps.setString(4, type);
            ps.setString(5, fileName);
            ps.setLong(6, fileSize);
            ps.executeUpdate();

            ResultSet keys = ps.getGeneratedKeys();
            if (keys.next()) return keys.getInt(1);

        } catch (SQLException e) {
            log.error("Lỗi lưu message: {}", e.getMessage());
        }
        return -1;
    }

    /**
     * Lấy lịch sử chat của một phòng (giới hạn 100 tin nhắn gần nhất).
     */
    public static List<Object[]> getMessagesByRoom(int roomId, int limit) {
        String sql = "SELECT m.id, m.sender_id, u.username, u.display_name, " +
                     "       m.encrypted_content, m.message_type, m.file_name, m.sent_at " +
                     "FROM messages m " +
                     "JOIN users u ON m.sender_id = u.id " +
                     "WHERE m.room_id = ? " +
                     "ORDER BY m.sent_at DESC LIMIT ?";

        List<Object[]> results = new ArrayList<>();
        try (Connection conn = DatabaseConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {

            ps.setInt(1, roomId);
            ps.setInt(2, limit);
            ResultSet rs = ps.executeQuery();

            while (rs.next()) {
                results.add(new Object[]{
                    rs.getInt("id"),
                    rs.getInt("sender_id"),
                    rs.getString("username"),
                    rs.getString("display_name"),
                    rs.getString("encrypted_content"),
                    rs.getString("message_type"),
                    rs.getString("file_name"),
                    rs.getTimestamp("sent_at")
                });
            }
        } catch (SQLException e) {
            log.error("Lỗi lấy messages: {}", e.getMessage());
        }
        return results;
    }

    /**
     * Xoá lịch sử chat theo phòng (admin).
     */
    public static boolean clearRoomHistory(int roomId) {
        String sql = "DELETE FROM messages WHERE room_id = ?";
        try (Connection conn = DatabaseConnection.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, roomId);
            ps.executeUpdate();
            return true;
        } catch (SQLException e) {
            log.error("Lỗi xoá lịch sử: {}", e.getMessage());
            return false;
        }
    }
}

File: SecureChat.Server/DAO/MessageDAO.cs
using System;
using System.Collections.Generic;
using MySqlConnector;

namespace SecureChat.Server.DAO
{
    public static class MessageDAO
    {
        /// <summary>
        /// Lưu tin nhắn mã hoá vào MySQL database.
        /// </summary>
        public static int SaveMessage(int roomId, int senderId, string encryptedContent, string type, string fileName, long fileSize)
        {
            // Đảm bảo phòng tồn tại trước khi lưu tin nhắn để tránh lỗi Foreign Key
            CreateRoomIfNotExist(roomId, $"Room_{roomId}");

            string sql = "INSERT INTO messages " +
                         "(room_id, sender_id, encrypted_content, iv, message_type, file_name, file_size) " +
                         "VALUES (@roomId, @senderId, @encryptedContent, '', @messageType, @fileName, @fileSize)";

            try
            {
                using (var conn = DatabaseConnection.GetConnection())
                using (var cmd = new MySqlCommand(sql, conn))
                {
                    cmd.Parameters.AddWithValue("@roomId", roomId);
                    cmd.Parameters.AddWithValue("@senderId", senderId);
                    cmd.Parameters.AddWithValue("@encryptedContent", encryptedContent);
                    cmd.Parameters.AddWithValue("@messageType", type);
                    cmd.Parameters.AddWithValue("@fileName", string.IsNullOrEmpty(fileName) ? DBNull.Value : fileName);
                    cmd.Parameters.AddWithValue("@fileSize", fileSize);
                    cmd.ExecuteNonQuery();
                    return (int)cmd.LastInsertedId;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[MessageDAO] Lỗi lưu message: {ex.Message}");
            }
            return -1;
        }

        /// <summary>
        /// Tạo phòng chat trong database nếu chưa tồn tại (tránh lỗi Foreign Key).
        /// </summary>
        public static void CreateRoomIfNotExist(int roomId, string roomName)
        {
            string sql = "INSERT IGNORE INTO chat_rooms (id, room_name, room_type) VALUES (@id, @name, 'PRIVATE')";
            try
            {
                using (var conn = DatabaseConnection.GetConnection())
                using (var cmd = new MySqlCommand(sql, conn))
                {
                    cmd.Parameters.AddWithValue("@id", roomId);
                    cmd.Parameters.AddWithValue("@name", roomName);
                    cmd.ExecuteNonQuery();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[MessageDAO] Lỗi tạo phòng chat {roomId}: {ex.Message}");
            }
        }

        /// <summary>
        /// Lấy lịch sử tin nhắn trong phòng (giới hạn số lượng).
        /// </summary>
        public static List<object[]> GetMessagesByRoom(int roomId, int limit)
        {
            var results = new List<object[]>();
            string sql = "SELECT m.id, m.sender_id, u.username, u.display_name, " +
                         "       m.encrypted_content, m.message_type, m.file_name, m.sent_at " +
                         "FROM messages m " +
                         "JOIN users u ON m.sender_id = u.id " +
                         "WHERE m.room_id = @roomId " +
                         "ORDER BY m.sent_at DESC LIMIT @limit";

            try
            {
                using (var conn = DatabaseConnection.GetConnection())
                using (var cmd = new MySqlCommand(sql, conn))
                {
                    cmd.Parameters.AddWithValue("@roomId", roomId);
                    cmd.Parameters.AddWithValue("@limit", limit);

                    using (var reader = cmd.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            results.Add(new object[]
                            {
                                reader.GetInt32("id"),
                                reader.GetInt32("sender_id"),
                                reader.GetString("username"),
                                reader.GetString("display_name"),
                                reader.GetString("encrypted_content"),
                                reader.GetString("message_type"),
                                reader.IsDBNull(reader.GetOrdinal("file_name")) ? "" : reader.GetString("file_name"),
                                reader.GetDateTime("sent_at")
                            });
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[MessageDAO] Lỗi lấy messages: {ex.Message}");
            }
            return results;
        }

        /// <summary>
        /// Xóa lịch sử phòng chat.
        /// </summary>
        public static bool ClearRoomHistory(int roomId)
        {
            string sql = "DELETE FROM messages WHERE room_id = @roomId";
            try
            {
                using (var conn = DatabaseConnection.GetConnection())
                using (var cmd = new MySqlCommand(sql, conn))
                {
                    cmd.Parameters.AddWithValue("@roomId", roomId);
                    cmd.ExecuteNonQuery();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[MessageDAO] Lỗi xóa lịch sử phòng {roomId}: {ex.Message}");
                return false;
            }
        }
    }
}
